Add docs for single operator multitenancy (#2581)
Azure/azure-service-operator: docs/hugo/content/introduction/multitenant-deployment.md

linktitle: Multitenancy
---

Currently, we support two types of multitenancy with Azure Service Operator (ASO): 

> ### Single operator multitenancy

Reviewer: Should we clarify that this is the default mode?

```suggestion
### Single operator multitenancy (default)
```

Reviewer: Or maybe:
```suggestion
### [Default] Single operator multitenancy
```

> Single operator deployed in the `azureserviceoperator-system` namespace with _a single set of global Azure credentials_,

Reviewer: minor: I think that you can remove the emphasis on "a single set of global Azure credentials" because now we actually support other creds.

Maybe say: 
```
Single operator deployed in the `azureserviceoperator-system` namespace.
This operator can be configured to manage resources with multiple different identities:
* Single global credential `aso-controller-settings` deployed as part of operator deployment.
* Per-namespace credential `aso-credential`. When this credential is created in a namespace, all resources in that namespace automatically use it instead of the global credential. This credential can be in a different tenant or subscription than the global credential.
* Per-resource credential, indicated by the `serviceoperator.azure.com/credential-from` annotation. This credential must be in the same namespace as the resource but can be in a different tenant or subscription than both the global credential and per-namespace credential.

When presented with multiple credential choices, the operator chooses the most specific one: per-resource takes precedence over per-namespace which takes precedence over global.
```

Reviewer: Possibly chop a bit from the above since you do cover some of it down below in the deployment section.

I do think it's at least clarifying the 3 flavors up here though (global/namespace/per-resource) as that doesn't really have that much to do with deployment per-se.

Thoughts?

Author: Nice! Yeah, makes sense to explain up here.

linktitle: Multitenancy
---

> Currently, we support two types of multitenancy with Azure Service Operator (ASO): 

Reviewer 2: Don't make readers scan down the document to find the second type of multitenancy. 

Either just give their names:

```suggestion
Currently, we support two types of multitenancy with Azure Service Operator (ASO): single operator and multiple operator.
```

or give one-line definitions:

```suggestion
Currently, we support two types of multitenancy with Azure Service Operator (ASO): 
* _single operator_ where a single ASO instance uses different credentials for different resources; and
* _multiple operator_ where an individual ASO instance is configured for each Kubernetes namespace.
```

> ### Single operator multitenancy (default)

Reviewer: Heading indentation levels don't really make sense here. These headings should be able the same level as the "XYZ deployment" ones below, so 2 `#`'s not 3.

Or else we could do:

```
## Single Operator multitenancy (default)
...

### Deployment
...

## Multiple operator multitenancy
...

### Deployment
...
```

Single operator deployed in the `azureserviceoperator-system` namespace.
This operator can be configured to manage resources with multiple different identities:

> * Single global credential `aso-controller-settings` deployed as part of operator deployment.

Reviewer 2: Which namespace does this go into?

Author: Made it clear b stating ' in operator's namespace'

> * Per-namespace credential `aso-credential`. When this credential is created in a namespace, all resources in that namespace automatically use it instead of the global credential. This credential can be in a different tenant or subscription than the global credential.
> * Per-resource credential, indicated by the `serviceoperator.azure.com/credential-from` annotation. This credential must be in the same namespace as the resource but can be in a different tenant or subscription than both the global credential and per-namespace credential.

Reviewer 2: Suggest using active voice, as though it's an instruction.

```suggestion
* Per-namespace credential `aso-credential`. Create this in a Kubernetes namespace to configure the credential used for all resources in that namespace, overriding the global credential. This credential can be in a different tenant or subscription than the global credential.
* Per-resource credential, indicated by the `serviceoperator.azure.com/credential-from` annotation. Create this annotation on each resource to configure the credential used for that resource. The secret containing the credential must be in the same Kubernetes namespace as the resource but may be in a different tenant or subscription than both the global credential and per-namespace credential.
```

When presented with multiple credential choices, the operator chooses the most specific one: per-resource takes precedence over per-namespace which takes precedence over global.

### Multiple operator multitenancy
Multiple operators deployed in different namespaces requires one deployment to handle webhooks (required because webhook configurations are cluster-level resources) and then a separate deployment for each tenant, each with its own credentials and set of namespaces that it watches for Azure resources.

ASO may also be deployed in a _multi-tenant_ configuration, enabling the use of separate credentials for managing resources in different Kubernetes namespaces.

## Single operator multitenancy deployment

To deploy the operator in single-operator multi-tenant mode:

1. Follow the normal ASO [installation](https://azure.github.io/azure-service-operator/#installation)

> 2. To use namespace scoped credential, create a credential secret named `aso-credential` in a namespace. Using this, all the resources in a namespace will use namespace scoped credential. 

Reviewer 2: ```suggestion
2. To use namespace scoped credential, create a credential secret named `aso-credential` in the desired namespace. Using this, all the resources in that namespace will use namespace scoped credential. 
```

secret/default-token-mqmpb       kubernetes.io/service-account-token   3      3d
```

### Helm Installation

To deploy the operator in multi-operator multi-tenant using helm is split into two parts:

1. **Cluster-wide operator installation**:
   ```
   helm repo add aso2 https://raw.githubusercontent.com/Azure/azure-service-operator/main/v2/charts
   ```

   ```

>    helm upgrade --install --devel aso2 aso2/azure-service-operator \
>    --create-namespace \
>    --namespace=azureserviceoperator-system \
>    --set azureSubscriptionID=$AZURE_SUBSCRIPTION_ID \
>    --set azureTenantID=$AZURE_TENANT_ID \
>    --set azureClientID=$AZURE_CLIENT_ID \
>    --set azureClientSecret=$AZURE_CLIENT_SECRET
>    --set multitenant.enabled=true
>    --set azureOperatorMode=webhooks

Reviewer 2: Suggest indenting after the first line.
```suggestion
   helm upgrade --install --devel aso2 aso2/azure-service-operator \
       --create-namespace \
       --namespace=azureserviceoperator-system \
       --set azureSubscriptionID=$AZURE_SUBSCRIPTION_ID \
       --set azureTenantID=$AZURE_TENANT_ID \
       --set azureClientID=$AZURE_CLIENT_ID \
       --set azureClientSecret=$AZURE_CLIENT_SECRET
       --set multitenant. Enabled=true
       --set azureOperatorMode=webhooks
```

   ```

2. **Per-tenant operator installation**:
   ```
   helm repo add aso2 https://raw.githubusercontent.com/Azure/azure-service-operator/main/v2/charts
   ```

   ```

>    helm upgrade --install --devel aso2 aso2/azure-service-operator \
>    --create-namespace \
>    --namespace=azureserviceoperator-system \
>    --set azureSubscriptionID=$AZURE_SUBSCRIPTION_ID \
>    --set azureTenantID=$AZURE_TENANT_ID \
>    --set azureClientID=$AZURE_CLIENT_ID \
>    --set azureClientSecret=$AZURE_CLIENT_SECRET
>    --set multitenant.enabled=true
>    --set azureOperatorMode=watchers

Reviewer 2: Ditto
```suggestion
   helm upgrade --install --devel aso2 aso2/azure-service-operator \
       --create-namespace \
       --namespace=azureserviceoperator-system \
       --set azureSubscriptionID=$AZURE_SUBSCRIPTION_ID \
       --set azureTenantID=$AZURE_TENANT_ID \
       --set azureClientID=$AZURE_CLIENT_ID \
       --set azureClientSecret=$AZURE_CLIENT_SECRET
       --set multitenant. Enabled=true
       --set azureOperatorMode=watchers
```
